sunpy/map/mapbase.py

```diff
@@ -1559,66 +1561,66 @@ def draw_limb(self, axes=None, **kwargs):
 
         return [circ]
 
+    @deprecate_positional_args
     @u.quantity_input
-    def draw_rectangle(self, bottom_left, width: u.deg, height: u.deg, axes=None, **kwargs):
+    def draw_rectangle(self, bottom_left, *, width: u.deg=None, height: u.deg=None,
+                       axes=None, top_right=None, **kwargs):
         """
         Draw a rectangle defined in world coordinates on the plot.
-
         Parameters
         ----------
-
-        bottom_left : `~astropy.coordinates.SkyCoord` or `~astropy.coordinates.BaseCoordinateFrame`
-            The bottom left corner of the rectangle.
-
-        width : `astropy.units.Quantity`
+        bottom_left : `astropy.units.Quantity` or `~astropy.coordinates.SkyCoord`
+            The bottom_left coordinate of the rectangle. If a `SkyCoord` it can
+            have shape ``(2,)`` and also define ``top_right``. If specifying
+            pixel coordinates it must be given as an `~astropy.units.Quantity`
+            object with units of `~astropy.units.pixel`.
+        top_right : `astropy.units.Quantity` or `~astropy.coordinates.SkyCoord`
+            The top_right coordinate of the rectangle. Can only be omitted if
+            ``bottom_left`` has shape ``(2,)``.
+        width : `astropy.units.Quantity` # passing it as a position argument is deprecated.
             The width of the rectangle.
-
-        height : `astropy.units.Quantity`
+        height : `astropy.units.Quantity` # passing it as a position argument is deprecated.
             The height of the rectangle.
-
         axes : `matplotlib.axes.Axes`
             The axes on which to plot the rectangle, defaults to the current
             axes.
-
         Returns
         -------
-
         rect : `list`
             A list containing the `~matplotlib.patches.Rectangle` object, after
             it has been added to ``axes``.
-
         Notes
         -----
-
         Extra keyword arguments to this function are passed through to the
         `~matplotlib.patches.Rectangle` instance.
-
         """
+        if any(x is None for x in (width, height)):
+            bottom_left, top_right = get_rectangle_coordinates(bottom_left,
+                                                               top_right=top_right,
+                                                               width=width,
+                                                               height=height)
+
+            width = np.abs(top_right.spherical.lon - bottom_left.spherical.lon).to(u.deg)
```

Comment:
This line may not be correct if the longitude crosses the longitude wrap.

Reply:
width and height have to be positive, so I think in case of a wrap, where we are getting a width of more than 180 degrees, we can do something like (180 - bigger longitude) + (180 + smaller longitude)
Because one of them has to be negative to get a wrap.
To prevent longitudes of more than 180 degrees in either direction, we take the modulus with 180.

Would this be enough?

Reply:
Widths larger than 180 degrees can be valid.  Do this:
```python
from astropy.coordinates import Longitude

width = Longitude(top_right.spherical.lon - bottom_left.spherical.lon)
```